### DISCUSSION TITLE
Running k3s unprivileged/rootless without systemd

### DISCUSSION THREAD

**Discussion by dev-zero** [k3s-io/k3s#7003] [Q&A] [answered]
I would like to run k3s with an unprivileged user in an environment where there is no systemd (resp. no systemd user session).
The error message I get is `cgroup v2 controllers are not delegated for rootless. Disabling cgroup.`, indicating this location:

https://github.com/k3s-io/k3s/blob/ee28c20b62d32aea012c96cb85d705f4c4db16cd/pkg/agent/containerd/config_linux.go#L49-L56

As a test: Giving the user permission to write `/sys/fs/cgroup` makes K3s start with `k3s server --rootless --snapshotter=fuse-overlayfs`, but fail later on with `"CreatePodSandbox for pod failed" err="rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: unable to apply cgroup configuration: failed to write 1092: write /sys/fs/cgroup/kubepods/besteffort/podce5ff44c-ce8e-4043-8c0d-9a85b053704d/67e42b6020bcb981d4411d8980e3b2bef301a17c8ff8f4bf959961a1e7ea3d0b/cgroup.procs: permission denied: unknown" pod="kube-system/helm-install-traefik-w2w74"`.

What are my options? I can do preparations (like creating a cgroup subgroup where the user has access and delegation is enabled) as root prior to running `k3s`, but the base cgroup path for the `cgroupfs` driver seems to be hardcoded (see above).
`unshare -mr` and then remounting `/sys/fs/cgroup` might be a way (and then not using `--rootless`)?
Or use RootlessKit to wrap the `k3s` invocation (and then not pass `--rootless` to k3s)?

Thanks in advance!

**Comment by brandond** [comment]
I would defer to @AkihiroSuda but I am not aware of any way to use rootless without the cgroup delegation managed by systemd.

**Reply by dev-zero** [reply]
Ok, thanks. There does not seem much missing, though.

I have tried with creating a group `k3s` with all controllers delegated and a subgroup `main`, then started `k3s` as a user and moved it to the `k3s/main` subgroup (moving as `root`). The error I end up with is:

```
E0227 10:25:57.868036      34 node_container_manager_linux.go:61] "Failed to create cgroup" err="mkdir /sys/fs/cgroup/kubepods: permission denied" cgroupName=[kubepods]
E0227 10:25:57.868079      34 kubelet.go:1466] "Failed to start ContainerManager" err="mkdir /sys/fs/cgroup/kubepods: permission denied"
```

Adding the argument `--kubelet-arg '--cgroup-root=k3s'` makes it progress until `runc`, where the error is:

```
ima-test:~$ E0227 10:24:47.957010      36 remote_runtime.go:176] "RunPodSandbox from runtime service failed" err="rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: unable to apply cgroup configuration: failed to write 5788: write /sys/fs/cgroup/k3s/kubepods/burstable/pod7fe42a81-69b7-4a48-afcc-b83d8446b238/d8bfe0a30340955ce006098e9a1fa1c59731ecb4721ab23ca17dc2239b98aead/cgroup.procs: no such file or directory: unknown"
E0227 10:24:47.957218      36 kuberuntime_sandbox.go:72] "Failed to create sandbox for pod" err="rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: unable to apply cgroup configuration: failed to write 5788: write /sys/fs/cgroup/k3s/kubepods/burstable/pod7fe42a81-69b7-4a48-afcc-b83d8446b238/d8bfe0a30340955ce006098e9a1fa1c59731ecb4721ab23ca17dc2239b98aead/cgroup.procs: no such file or directory: unknown" pod="kube-system/metrics-server-7b67f64457-n8fhc"
E0227 10:24:47.957332      36 kuberuntime_manager.go:782] "CreatePodSandbox for pod failed" err="rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: unable to apply cgroup configuration: failed to write 5788: write /sys/fs/cgroup/k3s/kubepods/burstable/pod7fe42a81-69b7-4a48-afcc-b83d8446b238/d8bfe0a30340955ce006098e9a1fa1c59731ecb4721ab23ca17dc2239b98aead/cgroup.procs: no such file or directory: unknown" pod="kube-system/metrics-server-7b67f64457-n8fhc"
E0227 10:24:47.957577      36 pod_workers.go:965] "Error syncing pod, skipping" err="failed to \"CreatePodSandbox\" for \"metrics-server-7b67f64457-n8fhc_kube-system(7fe42a81-69b7-4a48-afcc-b83d8446b238)\" with CreatePodSandboxError: \"Failed to create sandbox for pod \\\"metrics-server-7b67f64457-n8fhc_kube-system(7fe42a81-69b7-4a48-afcc-b83d8446b238)\\\": rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: unable to apply cgroup configuration: failed to write 5788: write /sys/fs/cgroup/k3s/kubepods/burstable/pod7fe42a81-69b7-4a48-afcc-b83d8446b238/d8bfe0a30340955ce006098e9a1fa1c59731ecb4721ab23ca17dc2239b98aead/cgroup.procs: no such file or directory: unknown\"" pod="kube-system/metrics-server-7b67f64457-n8fhc" podUID=7fe42a81-69b7-4a48-afcc-b83d8446b238
```
with the base `pod*` group being created but the subgroup `d8bfe0a30340955ce006098e9a1fa1c59731ecb4721ab23ca17dc2239b98aead` being missing:

```
lima-test:/sys/fs/cgroup/k3s# ls /sys/fs/cgroup/k3s/kubepods/burstable/pod7fe42a81-69b7-4a48-afcc-b83d8446b238/
cgroup.controllers        cgroup.type               cpuset.mems               io.max                    memory.stat
cgroup.events             cpu.idle                  cpuset.mems.effective     io.stat                   memory.swap.current
cgroup.freeze             cpu.max                   hugetlb.2MB.current       memory.current            memory.swap.events
cgroup.kill               cpu.max.burst             hugetlb.2MB.events        memory.events             memory.swap.high
cgroup.max.depth          cpu.stat                  hugetlb.2MB.events.local  memory.events.local       memory.swap.max
cgroup.max.descendants    cpu.weight                hugetlb.2MB.max           memory.high               pids.current
cgroup.procs              cpu.weight.nice           hugetlb.2MB.rsvd.current  memory.low                pids.events
cgroup.stat               cpuset.cpus               hugetlb.2MB.rsvd.max      memory.max                pids.max
cgroup.subtree_control    cpuset.cpus.effective     io.bfq.weight             memory.min
cgroup.threads            cpuset.cpus.partition     io.latency                memory.oom.group
```

My guess is then that I am on an inconsistent codepath with my setup, where that additional missing subdir should either not be there or needs to be created.

**Reply by dev-zero** [reply]
forgot to say: I patched out all the writable checks on `/sys/fs/cgroup`.

**Comment by dev-zero** [comment, marked as answer]
Success. I did not realise that k3s is taking care of moving itself to a subgroup. So, it is indeed possible.
What I did (essentially):

```bash
sudo mkdir /sys/fs/cgroup/k3s # or however you want to name it
sudo chown -R myuser: /sys/fs/cgroup/k3s
( kill -SIGSTOP $BASHPID; exec k3s server --rootless --snapshotter=fuse-overlayfs ) &
K3S_PID=$!
sudo echo $K3S_PID > /sys/fs/cgroup/k3s
kill -SIGCONT $K3S_PID
```

To contain the processes there needs to be an additional level of nesting, but the above shows the principle.
Also, if the user process starting k3s is already in a subgroup they can manipulate and has controllers enabled (something like `myuser/init` and then the user creates `myuser/k3s`), use of `sudo` above can be avoided.
Essentially replicating what `systemd` does automatically when using `systemd-run`.
Also note that compared to systemd, this solution will not cleanup the cgroups. In particular will k3s "convert" the previous "leaf" cgroup to a parent cgroup with controllers enabled, which will not permit to to attach the pid in the next run (see the limitations on delegated cgroups).

But with this I was able to run `k3s` rootless on Alpine.

Also, my previously mentioned patching out of the `/sys/fs/cgroup` writeability checks is unnecessary since once properly setup k3s/kubernetes is able to isolate the mounts properly such that the processes get a consistent and expected view of the cgroup filesystem and the checks succeed.
